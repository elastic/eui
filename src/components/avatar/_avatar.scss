.euiAvatar {
  flex-shrink: 0; // Ensures it never scales down below its intended size
  display: inline-flex;
  justify-content: center;
  align-items: center;
  background-size: cover;
  text-align: center;
  vertical-align: middle;
  overflow-x: hidden;
  font-weight: $euiFontWeightMedium; // Explicitly state so it doesn't get overridden by inheritance
}

.euiAvatar--user {
  @include innerBorder('dark', 50%, .05);
  border-radius: 50%;
}

.euiAvatar--space {
  @include innerBorder('dark', $euiBorderRadius, .05);
  border-radius: $euiBorderRadius;
}

.euiAvatar-isDisabled {
  cursor: not-allowed;
  filter: grayscale(100%);
}

.euiAvatar--plain {
  background-color: $euiColorEmptyShade;
}

// Modifiers for sizing.
$avatarSizing: (
  s: (
    size: $euiSizeL,
    font-size: $euiSizeM
  ),
  m: (
    size: $euiSizeXL,
    font-size: $euiSize*.9
  ),
  l: (
    size: $euiSizeXXL,
    font-size: $euiSizeL*.8
  ),
  xl: (
    size: ($euiSize * 4),
    font-size: $euiSizeXL*.8
  ),
);

@each $size, $map in $avatarSizing {
  .euiAvatar--#{$size} {
    @include size(map-get($map, 'size'));
    line-height: map-get($map, 'size');
    font-size: map-get($map, 'font-size');
  }
}
