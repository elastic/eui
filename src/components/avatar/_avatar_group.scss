.euiAvatarGroup {
  vertical-align: baseline;

  > .euiAvatar:not(:first-of-type) {
    margin-left: ($euiSizeXS * -1);
    box-shadow: 0 0 0 1px $euiColorGhost;
  }
}

.euiAvatarGroup__more {
  @include euiFontSizeS;
  font-weight: $euiFontWeightMedium;
  margin-left: $euiSizeXS;
}
