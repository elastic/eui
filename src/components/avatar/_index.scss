@import 'avatar';
@import 'avatar_group';
