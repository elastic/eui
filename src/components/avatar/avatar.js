import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { VISUALIZATION_COLORS } from '../../services/colors/visualization_colors';

const sizeToClassNameMap = {
  'none': null,
  's': 'euiAvatar--s',
  'm': 'euiAvatar--m',
  'l': 'euiAvatar--l',
  'xl': 'euiAvatar--xl',
};

export const SIZES = Object.keys(sizeToClassNameMap);

export const EuiAvatar = ({
  imageUrl,
  name,
  className,
  size,
  ...rest
}) => {
  const classes = classNames(
    'euiAvatar',
    sizeToClassNameMap[size],
    className
  );

  let optionalInitial;
  if (name && !imageUrl) {
    optionalInitial = (
      <span aria-hidden="true">{name.substring(0, 1)}</span>
    );
  }

  const assignedColor = VISUALIZATION_COLORS[Math.floor(name.length % VISUALIZATION_COLORS.length)];

  const avatarStyle = {
    backgroundImage: imageUrl ? 'url(' + imageUrl + ')' : 'none',
    backgroundColor: assignedColor,
  };

  return (
    <div
      className={classes}
      style={avatarStyle}
      aria-label={name}
      {...rest}
    >
      {optionalInitial}
    </div>
  );
};

EuiAvatar.propTypes = {
  className: PropTypes.string,
  imageUrl: PropTypes.string,
  name: PropTypes.string.isRequired,
  size: PropTypes.oneOf(SIZES),
};

EuiAvatar.defaultProps = {
  size: 'm',
};
