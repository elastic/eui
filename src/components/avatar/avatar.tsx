import React, { HTMLAttributes, FunctionComponent } from 'react';
import { CommonProps, keysOf } from '../common';
// import PropTypes from 'prop-types';
import classNames from 'classnames';

import { isColorDark, hexToRgb } from '../../services/color';
import { VISUALIZATION_COLORS } from '../../services';

const sizeToClassNameMap = {
  none: null,
  s: 'euiAvatar--s',
  m: 'euiAvatar--m',
  l: 'euiAvatar--l',
  xl: 'euiAvatar--xl',
};

export const SIZES = keysOf(sizeToClassNameMap);
export type EuiAvatarSize = keyof typeof sizeToClassNameMap;

const typeToClassNameMap = {
  space: 'euiAvatar--space',
  user: 'euiAvatar--user',
};

export const TYPES = keysOf(typeToClassNameMap);
export type EuiAvatarType = keyof typeof typeToClassNameMap;

export type EuiAvatarProps = HTMLAttributes<HTMLDivElement> &
  CommonProps & {
    /**
     * Full name of avatar for title attribute and calculating initial if not provided
     */
    name: string;

    /**
     * Accepts hex value `#FFFFFF`, `#000` otherwise a viz palette color will be assigned
     */
    color?: string;

    /**
     * Custom initials (max 2 characters).
     * By default will take the first character (of each word).
     */
    initials?: string;

    /**
     * Specify how many characters to show (max 2 allowed).
     * By default, will show based on number of words.
     */
    initialsLength?: 1 | 2;

    /**
     * The type of avatar to this is displaying
     */
    type?: EuiAvatarType;
    imageUrl?: string;
    size?: EuiAvatarSize;
  };

export const EuiAvatar: FunctionComponent<EuiAvatarProps> = ({
  className,
  color,
  imageUrl,
  initials,
  initialsLength,
  name,
  size = 'm',
  type = 'user',
  ...rest
}) => {
  const classes = classNames(
    'euiAvatar',
    sizeToClassNameMap[size],
    typeToClassNameMap[type],
    className
  );

  checkValidColor(color);
  checkValidInitials(initials);

  let optionalInitial;
  if (name && !imageUrl) {
    // Calculate the number of initials to show, maxing out at 2
    let calculatedInitialsLength = initials
      ? initials.split(' ').length
      : name.split(' ').length;
    calculatedInitialsLength =
      calculatedInitialsLength > 2 ? 2 : calculatedInitialsLength;

    // Check if initialsLength was passed and set to calculated, unless greater than 2
    if (initialsLength) {
      calculatedInitialsLength = initialsLength <= 2 ? initialsLength : 2;
    }

    let calculatedInitials;
    // A. Set to initials prop if exists (but trancate to 2 characters max unless length is supplied)
    if (initials) {
      calculatedInitials = initials.substring(0, calculatedInitialsLength);
    } else {
      if (name.trim() && name.split(' ').length > 1) {
        // B. If there are any spaces in the name, set to first letter of each word
        calculatedInitials = name.match(/\b(\w)/g);
        calculatedInitials =
          calculatedInitials &&
          calculatedInitials.join('').substring(0, calculatedInitialsLength);
      } else {
        // C. Set to the name's initials truncated based on calculated length
        calculatedInitials = name.substring(0, calculatedInitialsLength);
      }
    }

    optionalInitial = <span aria-hidden="true">{calculatedInitials}</span>;
  }

  const assignedColor =
    color ||
    VISUALIZATION_COLORS[Math.floor(name.length % VISUALIZATION_COLORS.length)];
  const textColor = isColorDark(...hexToRgb(assignedColor))
    ? '#FFFFFF'
    : '#000000';

  const avatarStyle = {
    backgroundImage: imageUrl ? `url(${imageUrl})` : 'none',
    backgroundColor: assignedColor,
    color: textColor,
  };

  return (
    <div
      className={classes}
      style={avatarStyle}
      aria-label={name}
      title={name}
      {...rest}>
      {optionalInitial}
    </div>
  );
};

// TODO: Migrate to a service
function checkValidColor(color: EuiAvatarProps['color']) {
  const validHex = color && /(^#[0-9A-F]{6}$)|(^#[0-9A-F]{3}$)/i.test(color);
  if (color && !validHex) {
    throw new Error(
      `EuiAvatar needs to pass a valid color. This can either be a three ` +
        `or six character hex value`
    );
  }
}

function checkValidInitials(initials: EuiAvatarProps['initials']) {
  // Must be a string of 1 or 2 characters
  if (initials && initials.length > 2) {
    // tslint:disable-next-line:no-console
    console.warn(
      `EuiAvatar only accepts a max of 2 characters for the initials as a string`
    );
  }
}
