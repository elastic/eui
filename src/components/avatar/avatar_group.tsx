import React, { HTMLAttributes, ReactNode } from 'react';
import { CommonProps } from '../common';
import classNames from 'classnames';
// import { EuiAvatar } from './avatar';

export type EuiAvatarGroupProps = HTMLAttributes<HTMLDivElement> &
  CommonProps & {
    /**
     * Array of avatars
     */
    avatars: ReactNode[];
    /**
     * How many avatars to show before displaying the more text
     */
    maxLength?: number;
  };

export const EuiAvatarGroup: React.FunctionComponent<EuiAvatarGroupProps> = ({
  className,
  avatars,
  maxLength = 3,
  ...rest
}) => {
  const classes = classNames('euiAvatarGroup', className);

  const visibleChildren = avatars.slice(0, maxLength).map(avatar => avatar);

  let moreNode;
  if (avatars.length > maxLength) {
    moreNode = (
      <span className="euiAvatarGroup__more">
        +{avatars.length - maxLength}
      </span>
    );
  }

  return (
    <div className={classes} {...rest}>
      {visibleChildren}
      {moreNode}
    </div>
  );
};
