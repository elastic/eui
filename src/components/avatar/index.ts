export { EuiAvatar } from './avatar';
export { EuiAvatarGroup } from './avatar_group';
