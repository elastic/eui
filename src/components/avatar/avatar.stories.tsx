/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import type { Meta, StoryObj } from '@storybook/react';

import { EuiAvatar, EuiAvatarProps } from './avatar';

const meta: Meta<EuiAvatarProps> = {
  title: 'EuiAvatar',
  component: EuiAvatar,
  argTypes: {
    initialsLength: {
      options: [undefined, 1, 2],
    },
    color: { control: 'text' },
    iconType: { control: 'text' },
    iconColor: { control: 'text' },
  },
  args: {
    // Component defaults
    casing: 'uppercase',
    size: 'm',
    type: 'user',
    isDisabled: false,
  },
};

export default meta;
type Story = StoryObj<EuiAvatarProps>;

export const Playground: Story = {
  args: {
    name: 'Hello World',
  },
};
