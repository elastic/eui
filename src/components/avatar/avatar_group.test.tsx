import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../test/required_props';

import { EuiAvatar } from './avatar';
import { EuiAvatarGroup } from './avatar_group';

describe('EuiAvatarGroup', () => {
  test('is rendered', () => {
    const component = render(
      <EuiAvatarGroup
        avatars={[
          <EuiAvatar key="1" name="Rafael" />,
          <EuiAvatar key="2" name="Donatello" />,
          <EuiAvatar key="3" name="Leornardo" color="#BD10E0" />,
          <EuiAvatar key="4" name="Michaelangelo" />,
        ]}
        {...requiredProps}
      />
    );

    expect(component).toMatchSnapshot();
  });
});
