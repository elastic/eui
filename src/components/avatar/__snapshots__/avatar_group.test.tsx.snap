// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiAvatarGroup is rendered 1`] = `
<div
  aria-label="aria-label"
  class="euiAvatarGroup testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    aria-label="Rafael"
    class="euiAvatar euiAvatar--m euiAvatar--user"
    style="background-image:none;background-color:#F98510;color:#000000"
    title="Rafael"
  >
    <span
      aria-hidden="true"
    >
      R
    </span>
  </div>
  <div
    aria-label="Donatello"
    class="euiAvatar euiAvatar--m euiAvatar--user"
    style="background-image:none;background-color:#920000;color:#FFFFFF"
    title="Donatello"
  >
    <span
      aria-hidden="true"
    >
      D
    </span>
  </div>
  <div
    aria-label="Leornardo"
    class="euiAvatar euiAvatar--m euiAvatar--user"
    style="background-image:none;background-color:#BD10E0;color:#FFFFFF"
    title="Leornardo"
  >
    <span
      aria-hidden="true"
    >
      L
    </span>
  </div>
  <span
    class="euiAvatarGroup__more"
  >
    +1
  </span>
</div>
`;
